osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsCommandExecutorIntegrationTest.java

```diff
@@ -0,0 +1,537 @@
+/**
+ * Copyright 2019 Smart Society Services B.V.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.periodicmeterreads;
+
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.openmuc.jdlms.AttributeAddress;
+import org.openmuc.jdlms.ObisCode;
+import org.openmuc.jdlms.SelectiveAccessDescription;
+import org.openmuc.jdlms.datatypes.CosemDateTime;
+import org.openmuc.jdlms.datatypes.DataObject;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.DlmsObjectConfigConfiguration;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.dlmsobjectconfig.DlmsObjectConfigService;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.stub.DlmsConnectionManagerStub;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.stub.DlmsConnectionStub;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.testutil.AttributeAddressAssert;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.utils.AmrProfileStatusCodeHelper;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.commands.utils.DlmsHelper;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.Protocol;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.PeriodTypeDto;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.PeriodicMeterReadsRequestDto;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.PeriodicMeterReadsResponseDto;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.PeriodicMeterReadsResponseItemDto;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(MockitoJUnitRunner.class)
+public class GetPeriodicMeterReadsCommandExecutorIntegrationTest {
```

Comment:
The name 'IntegrationTest' is confusing here because this test is run as a unit test by the maven-surefire-plugin. We also use the maven-failsafe-plugin which runs tests matching: `<include>%regex[.*integrationtests.*class]</include>` in the `integration-test` phase.

I think you named it such because this test is not a pure 'mockist' style unit tests. There is a 'classical' view on unit testing which would call this a real unit test, where the unit happens to be larger than a single class. I don't prefer such tests myself since the tests tend to become hard to read, repetitive and prone to miss edge cases. Other developers (read: @stefanermens) however are of a different opinion.

Reply:
I inherited this setup and did not touch the structure.

Reply:
The main reason for this test is to verify that when all the classes work together, the correct DLMS request is created. This is essential to test since a single small difference could cause the smart meter to send a wrong reply or no reply at all. This test tests all possible requests from the SMHE and compares the created requests with known good requests, which have been tested with real meters. So with a single test we can see that with a certain input (e.g. Daily meter reads, gas meter, SMR5) we get the right request. 

With pure 'mockist' style unit tests it would be very hard to prove the same. And when a code change is needed, the 'mockist' style unit test would need to be changed as well, which might lead to a wrong request being created, but not detected by the test.

About the naming we could have endless discussions. We did have a discussion last time and then choose to name it IntegrationTests. I don't mind what we choose, as long as we don't discuss it over and over again.